pkg/autoscaler/controller/laddercontroller/ladder_controller.go

```diff
@@ -0,0 +1,133 @@
+func NewLadderController() controller.Interface {
+	return &LadderController{}
+}
+
+type paramEntry [2]int
+
+type paramEntries []paramEntry
+
+func (entries paramEntries) Len() int {
+	return len(entries)
+}
+
+func (entries paramEntries) Less(i, j int) bool {
+	return entries[i][0] < entries[j][0]
+}
+
+func (entries paramEntries) Swap(i, j int) {
+	entries[i], entries[j] = entries[j], entries[i]
+}
+
+type ladderParams struct {
+	CoresToReplicasEntries paramEntries `json:"cores_to_replicas_entries"`
+	NodesToReplicasEntries paramEntries `json:"nodes_to_replicas_entries"`
+}
+
+func (c *LadderController) GetExpectedNumOfReplicas(paramsRaw, resourceVersion string, schedulableNodes, schedulableCores int32) (int32, error) {
+	// Only parse and rebuild the entries when the configmap changed
+	if resourceVersion != c.version {
+		glog.V(0).Infof("Detected ConfigMap version change (old: %s new: %s) - rebuilding lookup entries\n", c.version, resourceVersion)
+		params, err := parseParams([]byte(paramsRaw))
+		if err != nil {
+			return 0, fmt.Errorf("error parsing ladder params: %s", err)
+		}
+		sort.Sort(params.CoresToReplicasEntries)
+		sort.Sort(params.NodesToReplicasEntries)
+		c.params = params
+		c.version = resourceVersion
+	}
+
+	// Get the expected replicas for the currently schedulable nodes and cores
+	expReplicas := int32(c.getExpectedReplicasFromParams(int(schedulableNodes), int(schedulableCores)))
+
+	return expReplicas, nil
+}
+
+// parseParams Parse the params from JSON string
+func parseParams(data []byte) (params *ladderParams, err error) {
+	var p ladderParams
+	if err := json.Unmarshal(data, &p); err != nil {
+		return nil, fmt.Errorf("could not parse parameters (%s)", err)
+	}
+	for _, e := range p.CoresToReplicasEntries {
+		if len(e) != 2 {
+			return nil, fmt.Errorf("invalid element %v in cores_to_replicas_map", e)
+		}
+		if e[0] < 1 || e[1] < 1 {
+			return nil, fmt.Errorf("invalid negative values in entry %v in cores_to_replicas_map", e)
+		}
+	}
+	for _, e := range p.NodesToReplicasEntries {
+		if len(e) != 2 {
+			return nil, fmt.Errorf("invalid element %b in nodes_to_replicas_map", e)
+		}
+		if e[0] < 1 || e[1] < 1 {
+			return nil, fmt.Errorf("invalid negative values in entry %v in nodes_to_replicas_map", e)
+		}
+	}
+	return &p, nil
+}
+
+func (c *LadderController) getExpectedReplicasFromParams(schedulableNodes, schedulableCores int) int {
+	replicasFromCoreEntries := getExpectedReplicasFromEntries(schedulableCores, c.params.CoresToReplicasEntries)
+	replicasFromNodeEntries := getExpectedReplicasFromEntries(schedulableNodes, c.params.NodesToReplicasEntries)
+
+	// Returns the results which yields the most replicas
+	return int(math.Max(float64(replicasFromCoreEntries), float64(replicasFromNodeEntries)))
+}
+
+func getExpectedReplicasFromEntries(schedulableResources int, entries []paramEntry) int {
+	if len(entries) == 0 {
+		return 1
+	}
+	// Binary search for the corresponding replicas number
+	pos := sort.Search(
```

Comment:
I feel like both ways should be sufficient considering the entries number would not be large and the polling period is long enough?